test/xpu_api/numerics/numeric.ops/reduce/xpu_reduce_init_op.pass.cpp

```diff
@@ -0,0 +1,90 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// <numeric>
+// UNSUPPORTED: c++03, c++11, c++14
+// UNSUPPORTED: clang-8
+// UNSUPPORTED: gcc-9
+
+// Became constexpr in C++20
+// template<class InputIterator, class T>
+//   T reduce(InputIterator first, InputIterator last, T init);
+#include <oneapi/dpl/numeric>
+#include <cassert>
+#include <CL/sycl.hpp>
+
+#include "support/test_iterators.h"
+
+constexpr sycl::access::mode sycl_read = sycl::access::mode::read;
+constexpr sycl::access::mode sycl_write = sycl::access::mode::write;
+
+template <class T>
+class KernelTest;
+
+template <typename _T1, typename _T2>
+void
+ASSERT_EQUAL(_T1&& X, _T2&& Y)
+{
+    if (X != Y)
+        std::cout << "CHECK CORRECTNESS (STL WITH SYCL): fail (" << X << "," << Y << ")" << std::endl;
+}
+
+template <class Iter>
+void
+test(sycl::queue& deviceQueue)
+{
+    int input[] = {1, 2, 3, 4, 5, 6};
+    int output[8];
+    int ref[8] = {0, 1, 1, 2, 3, 6, 21, 2880};
+    sycl::range<1> numOfItems{6};
+    {
+        sycl::buffer<int, 1> buffer1(input, sycl::range<1>{6});
+        sycl::buffer<int, 1> buffer2(output, sycl::range<1>{8});
+        deviceQueue.submit(
+            [&](sycl::handler& cgh)
+            {
+                auto in = buffer1.get_access<sycl_read>(cgh);
+                auto out = buffer2.get_access<sycl_write>(cgh);
+                cgh.single_task<KernelTest<Iter>>(
+                    [=]()
+                    {
+                        out[0] = dpl::reduce(Iter(&in[0]), Iter(&in[0]), 0, std::plus<>());
+                        out[1] = dpl::reduce(Iter(&in[0]), Iter(&in[0]), 1, std::multiplies<>());
+                        out[2] = dpl::reduce(Iter(&in[0]), Iter(&in[0] + 1), 0, std::plus<>());
+                        out[3] = dpl::reduce(Iter(&in[0]), Iter(&in[0] + 1), 2, std::multiplies<>());
+                        out[4] = dpl::reduce(Iter(&in[0]), Iter(&in[0] + 2), 0, std::plus<>());
+                        out[5] = dpl::reduce(Iter(&in[0]), Iter(&in[0] + 2), 3, std::multiplies<>());
+                        out[6] = dpl::reduce(Iter(&in[0]), Iter(&in[0] + 6), 0, std::plus<>());
+                        out[7] = dpl::reduce(Iter(&in[0]), Iter(&in[0] + 6), 4, std::multiplies<>());
```

Comment:
what about custom user's lambdas? This is also applicable according to the C++ standard

Reply:
I will add a test for user's lambda.